src/compute/src/sink/persist_sink.rs

```diff
@@ -845,6 +851,24 @@ where
             correction_oks.advance_by(&persist_frontier);
             correction_errs.advance_by(&persist_frontier);
 
+            // Discard batch descriptions whose upper is already not beyond the
+            // persist frontier. Those have no chance of being applied to the
+            // shard succesfully.
+            let persist_upper = write.shared_upper();
+            in_flight_batches
+                .retain(|(lower, _upper), _cap| PartialOrder::less_equal(&persist_upper, lower));
+
+            if obey_read_only && read_only.borrow().clone() {
+                // We are not allowed to do writes, so go back to the beginning
+                // of the loop.
+                //
+                // We are bailing here and not earlier to make sure that we keep
+                // our corrections buffers up to date, which will potentially
+                // consolidate things out, and to make sure that we weed out
+                // batches that we can never apply.
+                continue;
```

Comment:
Nit: We might want to extract everything that follows this `if` into a separate function. I find it increasingly difficult to follow the control flow, and adding `continue`s doesn't make it simpler.

Reply:
That sounds good, yes! I'll leave that as follow-up work though because the PR is already big enough. 😅